lib/hyperswitch.js

```diff
@@ -241,39 +218,39 @@ HyperSwitch.prototype._checkInternalApiRequest = function(req) {
     }
 };
 
-HyperSwitch.prototype.request = function(req, options) {
-    if (req.method) {
-        req.method = req.method.toLowerCase();
-    }
-    return this._request(req, options);
-};
+HyperSwitch.prototype._createFilteredHandler = function(filters, handler, specInfo) {
+    var filterIdx = 0;
+    return function handlerWrapper(hyper, req) {
+        if (filters && filterIdx < filters.length) {
+            var filter = filters[filterIdx];
+            filterIdx++;
 
-HyperSwitch.prototype._wrapInAccessCheck = function(handlerPromise, match, childReq) {
-    var self = this;
-    // Don't need to check access restrictions on /sys requests,
-    // as these endpoints are internal, so can be accessed only
-    // within HyperSwitch. (See HyperSwitch.prototype.request) All required
-    // checks should be added and made at the root of the request chain,
-    // at /v1 level
-    if (!this._isSysRequest(childReq)
-            && match.permissions
-            && Array.isArray(match.permissions)
-            && match.permissions.length) {
-        self._authService = self._authService || new AuthService(match.value.specRoot);
-        self._authService.addRequirements(match.permissions);
-        if (childReq.method === 'get' || childReq.method === 'head') {
-            return P.all([
-                handlerPromise,
-                self._authService.checkPermissions(self, childReq)
-            ])
-            .then(function(res) { return res[0]; });
+            if (typeof filter === 'function') {
+                return filter(hyper, req, handlerWrapper, filter.options, specInfo);
+            }
+
+            if (filter.method
+                    && filter.method !== req.method
+                    && !(filter.method === 'get' && req.method === 'head')) {
+                return handlerWrapper(hyper, req);
+            }
+
+            return filter.filter(hyper, req, handlerWrapper, filter.options, specInfo);
         } else {
-            return self._authService.checkPermissions(self, childReq)
-            .then(function() { return handlerPromise; });
+            return P.method(handler)(hyper, req);
         }
-    } else {
-        return handlerPromise;
+    };
+};
+
+HyperSwitch.prototype.request = function(req, options) {
+    if (req.method) {
+        req.method = req.method.toLowerCase();
     }
+    return this._createFilteredHandler(
```

Comment:
Will we basically always have filters to set up, or is there a significant percentage of filter-less requests? If so, then it might be worth setting up a fast path for filter-less requests.